src/types/curve25519.rs

```diff
@@ -22,14 +26,101 @@ impl Curve25519PublicKey {
         })
     }
 
+    #[classmethod]
+    pub fn from_bytes(_cls: &Bound<'_, PyType>, bytes: &[u8]) -> Result<Self, KeyError> {
+        let key: &[u8; 32] = bytes.try_into().map_err(|_| {
+            KeyError::from(vodozemac::KeyError::InvalidKeyLength {
+                key_type: "Curve25519PublicKey",
+                expected_length: 32,
+                length: bytes.len(),
+            })
+        })?;
+
+        Ok(Self {
+            inner: vodozemac::Curve25519PublicKey::from_slice(key)?,
+        })
+    }
+
     pub fn to_base64(&self) -> String {
         self.inner.to_base64()
     }
 
+    pub fn to_bytes(&self) -> Py<PyBytes> {
+        convert_to_pybytes(self.inner.to_bytes().as_slice())
+    }
+
     #[classattr]
     const __hash__: Option<PyObject> = None;
 
     fn __eq__(&self, other: &Self) -> bool {
         self.inner == other.inner
     }
 }
+
+#[pyclass]
+#[derive(Clone)]
+pub struct Curve25519SecretKey {
+    pub(crate) inner: vodozemac::Curve25519SecretKey,
+}
+
+impl From<vodozemac::Curve25519SecretKey> for Curve25519SecretKey {
+    fn from(value: vodozemac::Curve25519SecretKey) -> Self {
+        Self { inner: value }
+    }
+}
+
+#[pymethods]
+impl Curve25519SecretKey {
+    #[new]
+    fn new() -> Self {
+        Self {
+            inner: vodozemac::Curve25519SecretKey::new(),
+        }
+    }
+
+    #[classmethod]
+    pub fn from_base64(_cls: &Bound<'_, PyType>, key: &str) -> Result<Self, KeyError> {
+        Self::from_bytes(
+            _cls,
+            base64_decode(key)
+                .map_err(|e| KeyError::from(vodozemac::KeyError::Base64Error(e)))?
+                .as_slice(),
+        )
+    }
+
+    #[classmethod]
+    pub fn from_bytes(_cls: &Bound<'_, PyType>, bytes: &[u8]) -> Result<Self, KeyError> {
+        let key: &[u8; 32] = bytes.try_into().map_err(|_| {
+            KeyError::from(vodozemac::KeyError::InvalidKeyLength {
+                key_type: "Curve25519SecretKey",
+                expected_length: 32,
+                length: bytes.len(),
+            })
+        })?;
+
+        Ok(Self {
+            inner: vodozemac::Curve25519SecretKey::from_slice(key),
+        })
+    }
+
+    pub fn to_base64(&self) -> String {
+        base64_encode(self.inner.to_bytes().as_slice())
+    }
+
+    pub fn to_bytes(&self) -> Py<PyBytes> {
+        convert_to_pybytes(self.inner.to_bytes().as_slice())
+    }
+
+    pub fn public_key(&self) -> Curve25519PublicKey {
+        Curve25519PublicKey {
+            inner: vodozemac::Curve25519PublicKey::from(&self.inner),
+        }
+    }
+
+    #[classattr]
+    const __hash__: Option<PyObject> = None;
+
+    fn __eq__(&self, other: &Self) -> bool {
```

Comment:
Is this needed for something? This potentially introduces a timing attack.

Reply:
I don't really know, I copy-pasted `Curve25519PublicKey` to be honest :innocent: 

It seems useless for both ? Should I also remove it from `Curve25519PublicKey` ?

Reply:
Comparing public keys is fine, you even might want to have a hashmap or set where the public keys act as hash keys.

For secret keys this isn't true.

Let's just remove it for the secret key.